### ISSUE TITLE
Does this napi_value have native data associated with it, and if so, what type is it?

### ISSUE THREAD

**Issue by gabrielschulhof** [nodejs/abi-stable-node#105]
Currently we have no way of distinguishing a `napi_value` that refers to an object that contains native data from a `napi_value` that refers to a plain object.

Furthermore, we have a fundamental distinction in the way V8 works vs. the way JSRT works (AFAICT):
In V8 you can only associate native data with an object if it is instantiated from a constructor for which a native data slot was set aside with `SetInternalFieldCount()`

whereas

In JSRT you can only associate native data with a plain object, because AFAICT you cannot tell JsCreateFunction that the objects you wish to construct from the resulting function are to have a native data slot.

So, currently calling `napi_unwrap()` on a `napi_value` created with `napi_new_instance()` from a constructor created with `napi_define_class()` will cause a segfault on JSRT, but not on V8. Interestingly `napi_wrap()` does not fail on JSRT when given an object created with `napi_new_instance()`. It would be nice to have a `JsCreateExternalFunction()` in JSRT meaning that it creates a constructor whose instances will be external objects rather than plain objects. We could use such a function to implement `napi_define_class()`.

I prefer the V8 way whereby an object can only have data associated with it if C++ controls the constructor because it provides not only a slot for storing data, but also type safety by way of `napi_instanceof()`. That is, given a `napi_value` for which `napi_get_type_of_value()` returns `napi_object`, I can know by way of `napi_instanceof()` both that the `napi_value` can hold native data, and that the native data is of a certain type.

This type safety does not really come up with ObjectWrap because the idea is that you only call methods on the object, rather than calling generic methods and passing in the object as a parameter. Nevertheless, this way of creating type safety is pretty weak, because in JS you can always change the context via `myWrappedObject.method.apply(somePlainObject)` and then one cannot rely on `napi_get_cb_this()` to make sure that the context is an object containing native data that is of the wrapped C++ class type.

In concrete terms this means that my NAPI implementation of iotivity-node will not work with ChakraCore because it will segfault.

**Comment by jasongin** [comment]
Related to this, @boingoing and I were discussing yesterday how `napi_wrap()` for JSRT currently uses the shim, via `node::ObjectWrap`. We would like all of the JSRT implementation to avoid the shim, but the JSRT APIs are currently inadequate there. We'd talked about the lack of ability to setup a callback for finalizing the external data, but making the constructor produce instances that can hold external data is a problem also.

> In JSRT you can only associate native data with a plain object

Actually I thought you can only associate native data with an objected created by `JsCreateExternalObject()` ?

> a `JsCreateExternalFunction()` in JSRT meaning that it creates a constructor whose instances will be external objects

I would expect an API with that name to create a function that *itself* can hold external data. I know that's not what we need.

Another related problem is `napi_define_class()` for JSRT currently tries to call `JsSetExternalData()` on the constructor returned from `JsCreateNamedFunction()` if a data pointer was passed in. I'm guessing that will fail, but we've probably never tested that case. (I wrote that code before I had a good understanding of how external data is handled in JSRT.)

**Comment by gabrielschulhof** [comment]
Yes, `JsCreateExternalObject()` is the right API, but it creates an object based on some internal constructor, not one that can be provided.

I'm not dead set on the name `JsCreateExternalFunction()`. Still, I couldn't come up with a succinct way of saying `JsCreateFunctionWhoseInstancesAreExternalObjects()`. The bottom line is I think we need this before we can truly start wrapping objects. And, I mean, the shim must provide V8-equivalent functionality, including this, so it must be there, right?

**Comment by gabrielschulhof** [comment]
Hmmm ... `JsConstructExternalObject()`?

**Comment by aruneshchandra** [comment]
Needed for #11

**Comment by jasongin** [comment]
@aruneshchandra, #11 should not be blocked by this, unless you consider use of the shim to be blocking. As I pointed out above, the JSRT implementation for object wrapping currently uses the shim, which works. I think this discussion is more about what we should do in the longer term. Or am I missing something @gabrielschulhof?

**Comment by gabrielschulhof** [comment]
AFAICT chakrashim does two things:

- In response to `Function::NewInstance` it calls `JsConstructValue()`, and uses `ObjectData` to store external data separately from the resulting object, since the resulting instance is not an external object.
- In response to `ObjectTemplate::NewInstance()` it calls `JsConstructExternalObject()` and sets the prototype using `JsSetPrototype()`.

This implementation acutally results in behaviour that is different from V8. The Object resulting from `Function::NewInstance()` is an instance of the constructor when examined with Javascript's `instanceof`, but the object created wih `ObjectTemplate::NewInstance()` is not. In contrast, both objects are considered instances of the function when examined with V8's implementation of `instanceof`.

In particular,

```C++
// hello.cc
#include <nan.h>

using namespace v8;

NAN_MODULE_INIT(Init) {
  Local<String> myClassName = Nan::New("MyClassName").ToLocalChecked();
  Local<String> displayName = Nan::New("displayName").ToLocalChecked();
  Local<String> theFunctionInstance =
    Nan::New("theFunctionInstance").ToLocalChecked();
  Local<String> theObjectInstance =
    Nan::New("theObjectInstance").ToLocalChecked();
  Local<String> theConstructor = Nan::New("theConstructor").ToLocalChecked();

  Local<FunctionTemplate> theTemplate =
    Nan::New<FunctionTemplate>();
  theTemplate->SetClassName(myClassName);
  Local<ObjectTemplate> instanceTemplate = theTemplate->InstanceTemplate();
  instanceTemplate->SetInternalFieldCount(1);
  Local<Function> theFunction = Nan::GetFunction(theTemplate).ToLocalChecked();
  Nan::Set(theFunction, displayName, myClassName);

  Local<Object> theNewFunctionInstance =
    Nan::NewInstance(theFunction).ToLocalChecked();
  Local<Object> theNewObjectInstance =
    Nan::NewInstance(instanceTemplate).ToLocalChecked();

  // Set the internal field pointer to something bogus
  Nan::SetInternalFieldPointer(theNewFunctionInstance, 0,
    &theNewFunctionInstance);
  Nan::SetInternalFieldPointer(theNewObjectInstance, 0,
    &theNewObjectInstance);

  Nan::Set(target, theFunctionInstance, theNewFunctionInstance);
  Nan::Set(target, theObjectInstance, theNewObjectInstance);
  Nan::Set(target, theConstructor, theFunction);
}

NODE_MODULE(hello, Init)
```
```JS
// index.js
var hello = require( "bindings" )( "hello" );

console.log( hello.theFunctionInstance instanceof hello.theConstructor );
console.log( hello.theObjectInstance instanceof hello.theConstructor );
```

the above addon will output
```
true
true
```
on V8, and
```
true
false
```
on ChakraShim.

So, to properly implement wrapping, it looks like we need to do what ChakraShim does in response to `Function::NewInstance()` and track external data ourselves, because in the implementation of `napi_new_instance()` objects are created with `JsConstructObject()` and thus cannot themselves hold external data.

... unless can haz new API `JsConstructExternalObject()` which combines the correct behaviour of `instanceof` created by `JsConstructObject()` with the ability to store external data provided by `JsCreateExternalObject()`. Implementing such an API would benefit both the shim and NAPI.

**Comment by jasongin** [comment]
Nice analysis @gabrielschulhof .

I see [where the shim intercepts the function callback](https://github.com/nodejs/node-chakracore/blob/xplat/deps/chakrashim/src/v8functiontemplate.cc#L109-L119). If it's a constructor call, it *discards* the `this` argument (which was just constructed, and would be a proper instance of the constructor), and replaces it with a new object created by `JsCreateExternalObject()`.

While the shim could fix that `instanceof` bug by using a hypothetical `JsConstructExternalObject()` API there instead, and NAPI could use that also for wrapping without the shim, I'm not sure that would the be best design. I think it would be better if the `this` argument received by the ChakraCore function callback was already an external object in that case. So I prefer `JsCreateFunctionWhoseInstancesAreExternalObjects()` as suggested above. Or for a shorter name, maybe `JsCreateExternalConstructor()` ?

**Comment by jasongin** [comment]
I looked into a possible `JsCreateFunctionWhoseInstancesAreExternalObjects()` API for ChakraCore, and found it was going to be very complicated to implement. Then I considered the other alternative, `JsConstructExternalObject()`. The only thing different that would need to do is use the correct prototype when creating the external object. Then I realized it's possible to set the prototype afterward. So, now I have a working non-shim `napi_wrap` implementation that has correct `instanceof` results, without requiring any new JSRT APIs. It relies on injecting an external object via the constructor callback adapter, similar to the shim, but with the correct prototype:

```C++
JsValueRef CALLBACK CallbackWrapper(JsValueRef callee, bool isConstructCall, JsValueRef *arguments, unsigned short argumentCount, void *callbackState) {
  CallbackInfo cbInfo;
  cbInfo.thisArg = reinterpret_cast<napi_value>(arguments[0]);
  cbInfo.isConstructCall = isConstructCall;

  if (isConstructCall) {
    // For constructor callbacks, replace the the 'this' arg with a new external object,
    // to support wrapping a native object in the external object.
    JsValueRef externalThis;
    if (JsNoError == JsCreateExternalObject(
        nullptr, jsrtimpl::ExternalData::Finalize, &externalThis)) {

      // Copy the prototype from the default 'this' arg to the new 'external' this arg.
      if (arguments[0] != nullptr) {
        JsValueRef thisPrototype;
        if (JsNoError == JsGetPrototype(arguments[0], &thisPrototype)) {
          JsSetPrototype(externalThis, thisPrototype);
        }
      }

      cbInfo.thisArg = reinterpret_cast<napi_value>(externalThis);
    }
  }

  ...
```

I will include this in the ChakraCore implementation of the reference and external improvements that I'm working on.

**Comment by gabrielschulhof** [comment]
Yeah, it's super-important that the resulting object be indistinguishable from what would've been constructed via `JsConstructObject()`, including the fact that it gets passed to the constructor as its `this` argument.

**Comment by jasongin** [comment]
This is resolved by the introduction of a `napi_external` type and refactoring of `napi_wrap` implementation, done now in all branches.
